### PR TITLE
Add integration test for self-eviction

### PR DESCRIPTION
**Pull request by mennopruijssers**
This test validates self-eviction by:
- stop gossiping in the whole cluster
- send a `SIGTERM ` to SUT
- validate if the SUT declares itself as faulty by pinging other members with the change

### CODE REVIEW

File: test/test-coordinator.js [hunk 1]
```diff
@@ -303,7 +303,7 @@ TestCoordinator.prototype.shutdown = function shutdown() {
     if(this.adminChannel.topChannel.destroyed === false) {
         this.adminChannel.topChannel.close();
     }
-    this.sutProc.kill();
+    this.sutProc.kill('SIGKILL');
```
Comment (mennopruijssers) [inline test/test-coordinator.js]: This change is necessary to force kill the the sut when the test is completed.

File: test/self-eviction-tests.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+var test2 = require('./test-util').test2;
+
+
+test2('self eviction changes state to faulty and pings members', getClusterSizes(2), 2000000, prepareCluster(function(t, tc, n) {
+    return [
+        // disable pinging from fake nodes
+        dsl.disableAllNodesPing(t, tc),
+
+        // stop gossiping on SUT
+        dsl.stopGossip(t, tc),
```
Comment (btromanova) [inline test/self-eviction-tests.js]: Can you pls add a comment to justify stopGossip and disableAllNodesPing?

File: test/self-eviction-tests.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        var pings = _.filter(list, {
+            type: events.Types.Ping,
+            direction: 'request'
+        });
+        if (pings.length === 0) {
+            return cb(list);
+        }
+
+        var maxPings = Math.ceil(n * 0.4);
+        t.ok(pings.length <= maxPings, 'number of pings does not exceed 40% of cluster');
```
Comment (btromanova) [inline test/self-eviction-tests.js]: Do we need also to check that at least some number of pings are sent?

Comment (mennopruijssers) [inline test/self-eviction-tests.js]: Good catch. I changed line 62 to `return cb(null)`. That'll make sure the assertValidPings method is rerun if there are 0 pings.

**Comment by btromanova** [comment]
lgtm

File: test/self-eviction-tests.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+}));
+
+/**
+ * Validate the pings. Assert that:
+ * - the number of pings doesn't exceed 40% of the cluster
+ * - the pings are all to different members
+ * - the pings should declare the SUT faulty and originate from the SUT itself
+ *
+ * @param t the current test suite
+ * @param tc the test coordinated
```
Comment (thanodnl) [inline test/self-eviction-tests.js]: the param for `n` is missing.

File: test/self-eviction-tests.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * @return {assertValidPings} the assert functions
+ */
+function assertValidPings(t, tc, n) {
+    return function assertValidPings(list, cb) {
+        var pings = _.filter(list, {
+            type: events.Types.Ping,
+            direction: 'request'
+        });
+        if (pings.length === 0) {
```
Comment (thanodnl) [inline test/self-eviction-tests.js]: If I remember correctly this assertion function will be called for every incoming packet directly. This branch will not be hit as soon as the first faulty declaration is received. causing all subsequent  assertions to only be done on the first ping.

I think we should change the `pings.length === 0` to `pings.length !== expectedFaultyPings` to make sure we assert on the correct set of pings.

Comment (mennopruijssers) [inline test/self-eviction-tests.js]: It's called after `waitForGracefulShutdown` (which waits until the process is exited).

**Pull request merged**